src/main/java/org/opensearch/knn/index/query/nativelib/NativeEngineKnnVectorQuery.java

```diff
@@ -87,32 +89,29 @@ public Weight createWeight(IndexSearcher indexSearcher, ScoreMode scoreMode, flo
         return createDocAndScoreQuery(reader, topK).createWeight(indexSearcher, scoreMode, boost);
     }
 
-    private List<Map<Integer, Float>> doSearch(
+    private List<Map.Entry<LeafReaderContext, Map<Integer, Float>>> doSearch(
         final IndexSearcher indexSearcher,
         List<LeafReaderContext> leafReaderContexts,
         KNNWeight knnWeight,
         int k
     ) throws IOException {
-        List<Callable<Map<Integer, Float>>> tasks = new ArrayList<>(leafReaderContexts.size());
+        List<Callable<Map.Entry<LeafReaderContext, Map<Integer, Float>>>> tasks = new ArrayList<>(leafReaderContexts.size());
         for (LeafReaderContext leafReaderContext : leafReaderContexts) {
             tasks.add(() -> searchLeaf(leafReaderContext, knnWeight, k));
         }
         return indexSearcher.getTaskExecutor().invokeAll(tasks);
     }
 
-    private List<Map<Integer, Float>> doRescore(
+    private List<Map.Entry<LeafReaderContext, Map<Integer, Float>>> doRescore(
```

Comment:
This data structure is getting complex, can we look into simplifying it in favor of readability?